### DISCUSSION TITLE
LightweightGraph null construction

### DISCUSSION THREAD

**Discussion by johnmcgowan1** [lwtnn/lwtnn#147] [Q&A] [answered]
Hi lwtnn experts,

A quick question about using LightweightGraph in an ATLAS analysis. 

How can I make an instance of LightweightGraph without explicitly passing the ifstream. Should LightweightGraph have a null constructor? I'd like to declare a couple of lightweight graphs in the header of my analysis class, load the jsons just once in my initialize function, and then of course evaluate in my event loop. However my understanding is that because there is no null constructor, I can't declare a LightweightGraph in the header file. I need an ifstream for the json which I can also not put in the header. The example on the github readme declares and uses a LightweightGraph in one function so there is never any issue of scope. Is there a simple way around this that I just don't know of?

Thanks in advance for any discussion.

Best,
John

**Comment by dantrim** [comment, marked as answer]
Hi @johnmcgowan1,

In your analysis class can you simply make the `LightweightGraph` data members pointer types? For example:

```c++
class YourAnalysisClass
{
   ...
   private :
   ...
       std::unique_ptr<LightweightGraph> m_graph0;
       std::unique_ptr<LightweightGraph> m_graph1;
   ...
}
```

And then in the your constructor/initialization do something like:
```c++
YourAnalysisClass::initialize() {
     std::ifstream file0 = ....; // get the input JSON file
     std::ifstream file1 = ....; // get the other input JSON file
     m_graph0 = std::make_unique<LightweightGraph>(file0);
     m_graph1 = std::make_unique<LightweightGraph>(file1);
}
```

**Reply by dguest** [reply]
Just to add: the lack of a null constructor is by design. I want to make it as difficult as possible to call a misconfigured object. As @dantrim says it's pretty easy to work around this using pointers. I believe [`std::optional`](https://en.cppreference.com/w/cpp/utility/optional) should also allow you to do this.

**Reply by johnmcgowan1** [reply]
Hi @dantrim and @dguest ,

Thank you for your quick responses! The lack of null constructor makes sense and making the lightweightgraphs pointers like in the example you gave solved my problem.

Thanks again!

John

**Reply by dguest** [reply]
The other thing I should mention (sort of specific to ATLAS code) is that the error reporting is via exceptions. `lwt::parse_json_graph` can throw a few different ways from the underlying boost libraries, but all of them should inherit from `std::exception`. The Constructor itself will throw [`lwt::NNConfigurationException`][1] which, again, inherits from `std::exception` via `std::logic_error`. 

So if you wrap the initialization in a `try { ... } catch (std::exception& ex) { ... }` block you can return any configuration errors via status codes from your `initialize()` method (as is the ATLAS standard).

[1]: https://github.com/lwtnn/lwtnn/blob/master/include/lwtnn/Exceptions.hh#L17

**Reply by johnmcgowan1** [reply]
Thanks again @dguest ! I've wrapped my initializations in a try/catch block now.

One more question about running in the ATLAS environment. My code seems to be running fine with the lightweightgraph and it makes predictions (all ones, but I'm pretty sure this is because I haven't properly configured the standardization to be part of the network so its done on each event as it was during the training rather than being an indication that the lightweightgraph isn't working)

The only thing I do before running is set up AnalysisBase,21.2.164. But it looks to me that lwtnn isn't in this release. I'm surprised that it seems to work. TStreamerInfo does complain a bit:
```TStreamerInfo::Build      ERROR   VBSWyAnalysis, unknown type: lwt::LightweightGraph even_nn```

Should I expect lwtnn to work after doing this minimal setup? If so, how can I get TStreamer to recognize the type?

**Reply by dguest** [reply]
Why don't you think lwtnn in the release? It definitely should be (it's been there for quite a few years) and if you don't get compilation errors it sounds like CMake is finding it. Just for reference, it's [an external package](https://gitlab.cern.ch/atlas/atlasexternals/-/tree/1.0/External/lwtnn) (the 1.0 branch of externals corresponds to r21).

The streamer issues are more a ROOT thing. There's some magic syntax to avoid them (maybe something like `//!` after member data in a class), but this isn't something I'm an expert on. I'd ask on some more specific ROOT related forum.

**Reply by johnmcgowan1** [reply]
Sorry @dguest. I was looking in the wrong place... the athena git repository. Thank you for the link pointing it out.

I won't worry too much about the TStreamer error for the time being. The lightweight graph computation gives me a segfault when running on the grid which did not happen at all running locally. This is probably something I should follow up with distributed computing experts though. Locally, everything is running fine.

Thanks all again for your help!
